src/data_store/generic_data_store.cpp

```diff
@@ -293,4 +311,57 @@ void generic_data_store::init_minibatch() {
   }
 }
 
+std::pair<std::string, std::string> generic_data_store::get_pathname_and_prefix(std::string s) {
+  int num_slash = std::count(s.begin(), s.end(), '/');
+  if (num_slash < 1 || s.back() == '/') {
+    std::stringstream err;
+    err << __FILE__ << " " << __LINE__ << " :: "
+        << "<string> must be of the form: <pathname>/prefix";
+    throw lbann_exception(err.str());
+  }
+
+  size_t j = s.rfind('/');
+  std::string prefix = s.substr(j+1);
+  std::string pathname = s.substr(0, j);
+  return std::make_pair(prefix, pathname);
+}
+
+void generic_data_store::create_dirs(std::string s) {
+  if (s.back() != '/') {
+    s += '/';
+  }
+  size_t i = s.find('/', 1);
+  while (i != std::string::npos) {
+    std::string s2 = s.substr(0, i);
+    const int dir_err = mkdir(s2.c_str(), S_IRWXU | S_IRWXG | S_IROTH | S_IXOTH);
```

Comment:
The utility function make_dir() in src/io/file_io.cpp makes sure that only the master rank creates a directory and synchronizes. However, if there are large number of directories to create, this function is not efficient. 
In that case, you can use this function (or the similar existing function create_dir() in src/utils/file_utils.cpp) to create directories, but have only the master create directories and make sure to use barrier after that.

Reply:
Didn't know about that; I'll check it out. Thanks.